File: packages/storage/src/drive-http.ts
/**
 * Wire-level port the GoogleDriveAdapter talks to. The Drive REST API
 * has a handful of distinct surfaces (search, simple upload, resumable
 * upload start, resumable upload chunk, delete, get) — modelling them
 * here lets the test harness mock each one without an HTTP layer. The
 * production binding (M06 web shell, M12 Android shell) wires a real
 * `fetch` to each method.
 *
 * Bearer-token plumbing is the adapter's responsibility: it asks the
 * caller (AuthProvider.providerAccessToken("GOOGLE_DRIVE")) for a fresh
 * token on every call. The port stays auth-agnostic.
 */

export interface DriveListMatch {
  readonly id: string;
  readonly name: string;
  readonly mimeType: string;
  readonly size: number;
}

export interface ResumableUploadHandle {
  /**
   * URL returned by Drive's "start resumable upload" endpoint. The
   * adapter PUTs to this URL with `Content-Range` headers.
   */
  readonly sessionUri: string;
}

export interface DriveHttpClient {
  /**
   * Find a folder by name under [parentId] (or root, when null) owned by
   * the app. Returns the first match's id or null. Used by
   * `ensureAppFolder()`.
   */
  findAppFolder(token: string, parentId: string | null, name: string): Promise<string | null>;

  /** Create a folder; returns the new folder's id. */
  createFolder(token: string, parentId: string | null, name: string): Promise<string>;

  /** Simple (multipart) upload for files under the resumable threshold. */
  simpleUpload(
    token: string,
    folderId: string,
    name: string,
    mimeType: string,
    bytes: Uint8Array,
    folderHint: string | undefined,
  ): Promise<DriveListMatch>;

  /** Start a resumable upload session; returns the session URI. */
  startResumableUpload(
    token: string,
    folderId: string,
    name: string,
    mimeType: string,
    sizeBytes: number,
    folderHint: string | undefined,
  ): Promise<ResumableUploadHandle>;

  /**
   * PUT a chunk to the session URI. Returns the new file metadata when
   * the upload completes; returns null when the session is awaiting
   * more chunks.
   */
  putChunk(
    sessionUri: string,
    chunk: Uint8Array,
    rangeStart: number,
    totalSize: number,
  ): Promise<DriveListMatch | null>;

  /** Drive `files.get` returning size + name (used by the sign helper). */
  getMetadata(token: string, externalId: string): Promise<DriveListMatch>;

  delete(token: string, externalId: string): Promise<void>;
}

File: packages/android/storage-android/src/main/kotlin/app/notes/storage/DriveHttp.kt
package app.notes.storage

import java.io.InputStream

/**
 * Wire-level port the GoogleDriveAdapter talks to. Mirrors the TS twin's
 * `DriveHttpClient`. The Android shell (M12) wires a Ktor implementation;
 * tests use a scripted fake.
 */
interface DriveHttp {
    suspend fun findAppFolder(token: String, parentId: String?, name: String): String?
    suspend fun createFolder(token: String, parentId: String?, name: String): String
    suspend fun simpleUpload(
        token: String,
        folderId: String,
        name: String,
        mimeType: String,
        bytes: ByteArray,
        folderHint: String?,
    ): DriveFileMeta
    suspend fun startResumableUpload(
        token: String,
        folderId: String,
        name: String,
        mimeType: String,
        sizeBytes: Long,
        folderHint: String?,
    ): ResumableUploadHandle
    /**
     * PUT a chunk to the session URI. Returns the new file metadata when
     * the upload completes; returns null when the session is awaiting
     * more chunks.
     */
    suspend fun putChunk(
        sessionUri: String,
        chunk: ByteArray,
        rangeStart: Long,
        totalSize: Long,
    ): DriveFileMeta?
    suspend fun getMetadata(token: String, externalId: String): DriveFileMeta
    suspend fun delete(token: String, externalId: String)
}

data class DriveFileMeta(
    val id: String,
    val name: String,
    val mimeType: String,
    val size: Long,
)

data class ResumableUploadHandle(val sessionUri: String) {
    init { require(sessionUri.isNotBlank()) { "ResumableUploadHandle.sessionUri cannot be empty" } }
}

/** Token-supplier alias the adapter accepts. The Android shell wires this
 * to `AuthProvider.providerAccessToken(GoogleDrive)`. */
typealias TokenProvider = suspend () -> String?

/** Async chunk source the adapter pulls from during a resumable upload. */
fun interface ChunkSource {
    suspend fun nextChunk(): ByteArray?
}

/** Adapter from a blocking InputStream → ChunkSource for callers
 * that already have a stream-shaped FileInput. */
class InputStreamChunkSource(
    private val stream: InputStream,
    private val chunkBytes: Int = DEFAULT_CHUNK_BYTES,
) : ChunkSource {
    override suspend fun nextChunk(): ByteArray? {
        val buffer = ByteArray(chunkBytes)
        val read = stream.read(buffer)
        return when {
            read < 0 -> null
            read == buffer.size -> buffer
            else -> buffer.copyOf(read)
        }
    }

    companion object {
        const val DEFAULT_CHUNK_BYTES: Int = 8 * 1024 * 1024
    }
}
